lua/wire/client/cl_wirelib.lua

```diff
@@ -102,108 +111,97 @@ end
 
 
 function Wire_DrawTracerBeam( ent, beam_num, hilight, beam_length )
-	local beam_length = beam_length or ent:GetBeamLength(beam_num)
-	if (beam_length > 0) then
-
-		local x, y = 0, 0
-		if (ent.GetSkewX and ent.GetSkewY) then
-			x, y = ent:GetSkewX(beam_num), ent:GetSkewY(beam_num)
-		end
-
-		local start, ang = ent:GetPos(), ent:GetAngles()
-
-		if (ent.ls != start or ent.la != ang or ent.ll != beam_length or ent.lx != x or ent.ly != y) then
-			ent.ls, ent.la = start, ang
-
-			if (ent.ll != beam_length or ent.lx != x or ent.ly != y) then
-				ent.ll, ent.lx, ent.ly = beam_length, x, y
-
-				if (x == 0 and y == 0) then
-					ent.endpos = start + (ent:GetUp() * beam_length)
-				else
-					local skew = Vector(x, y, 1)
-					skew = skew*(beam_length/skew:Length())
-					local beam_x = ent:GetRight()*skew.x
-					local beam_y = ent:GetForward()*skew.y
-					local beam_z = ent:GetUp()*skew.z
-					ent.endpos = start + beam_x + beam_y + beam_z
-				end
-				ent.ExtraRBoxPoints = ent.ExtraRBoxPoints or {}
-				ent.ExtraRBoxPoints[beam_num] = ent:WorldToLocal(ent.endpos)
-			else
-				ent.endpos = ent:LocalToWorld(ent.ExtraRBoxPoints[beam_num])
-			end
-		end
-
-		local trace = {}
-		trace.start = start
-		trace.endpos = ent.endpos
-		trace.filter = { ent }
-		if ent:GetNWBool("TraceWater") then trace.mask = MASK_ALL end
-		trace = util.TraceLine(trace)
-
-		render.SetMaterial(beam_mat)
-		render.DrawBeam(start, trace.HitPos, 6, 0, 10, ent:GetColor())
-		if (hilight) then
-			render.SetMaterial(beamhi_mat)
-			render.DrawBeam(start, trace.HitPos, 6, 0, 10, Color(255,255,255,255))
+	local beam_length 	= 	beam_length or ent:GetBeamLength(beam_num)
+	if beam_length == 0 then return end
+	local pos 		= 	ent:GetPos()
+	local trace 		= 	{}
+	
+	if ent.GetTarget and ( ent:GetTarget().X ~= 0 or ent:GetTarget().Y ~= 0 or ent:GetTarget().Z ~= 0 ) then
+		trace.endpos 	= 	pos + ( ent:GetTarget() - pos ):GetNormalized()*beam_length
+		if trace.endpos[1] ~= trace.endpos[1] then trace.endpos = pos+Vector(ent:GetBeamLength(), 0, 0) end
+	elseif (ent.GetSkewX and ent.GetSkewY) then
+		local x, y = ent:GetSkewX(beam_num), ent:GetSkewY(beam_num)
+		if x ~= 0 or y ~= 0 then
+			local skew 		= 	Vector(x, y, 1)
+			skew 			= 	skew*(beam_length/skew:Length())
+			local beam_x 		= 	ent:GetRight()*skew.x
+			local beam_y 		= 	ent:GetForward()*skew.y
+			local beam_z 		= 	ent:GetUp()*skew.z
+			trace.endpos 		= 	pos + beam_x + beam_y + beam_z
+		else
+			trace.endpos 	= 	pos + ent:GetUp()*beam_length
 		end
+	else
+		trace.endpos 		= 	pos + ent:GetUp()*beam_length
+	end
+	
+	trace.start 		= 	pos
+	trace.filter 		= 	{ ent }
+	if ent:GetNWBool("TraceWater") then trace.mask = MASK_ALL end
+	trace 			= 	util.TraceLine(trace)
+
+	
+	if hilight then
+		render.SetMaterial(BeamMatHR)
+		render.DrawBeam(pos, trace.HitPos, 6, 0, 10, Color(255,255,255,255))
+	else
+		render.SetMaterial(BeamMat)
+		render.DrawBeam(pos, trace.HitPos, 6, 0, 10, ent:GetColor())
```

Comment:
The old code was correct, it draws two beams if hilight. This code only draws one beam.